src/Pirouette/SMT/Constraints.hs

```diff
@@ -1,167 +1,276 @@
 {-# LANGUAGE FlexibleContexts #-}
 {-# LANGUAGE FlexibleInstances #-}
 {-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE ScopedTypeVariables #-}
 {-# LANGUAGE TypeFamilies #-}
 {-# LANGUAGE UndecidableInstances #-}
+{-# OPTIONS_GHC -Wno-unrecognised-pragmas #-}
+
+{-# HLINT ignore "Collapse lambdas" #-}
 
 -- | Constraints that we can translate to SMT
 module Pirouette.SMT.Constraints where
 
+import Control.Applicative
+import Control.Monad
+import Data.Default
 import Data.Either
-import Data.List (intersperse)
-import Data.Map (Map)
+import qualified Data.Map as M
 import qualified Data.Set as S
 import Pirouette.Monad
 import Pirouette.SMT.Base
 import Pirouette.SMT.FromTerm
 import Pirouette.Term.Syntax
+import qualified Pirouette.Term.Syntax.SystemF as SystF
 import Prettyprinter hiding (Pretty (..))
 import qualified PureSMT
 
--- TODO: this module should probably be refactored somewhere;
--- I'm not entirely onboard with the 'translateData' funct as it is;
---   a) maybe we should pass the env when translating arbitrary expressions
---   b) maybe we should even keep the env in another reader layer on SolverT...
---
--- Its somthing to think about later, but for now this will do.
-
--- | Bindings from names to types (for the assign constraints)
-type Env lang = Map Name (Type lang)
-
--- | Constraints of a path during symbolic execution
--- We would like to have
--- type Constraint = Bot | And (Map Name [PirTerm]) [(Fuel, PirTerm, PirTerm))]
--- It is isomorphic to our current type, but with better access time to the variable assignements.
--- The 2 kinds of equalities are quite different,
--- one represents the complete execution of a branch, which leads to an application which cannot be further inlined
--- (Because it is a builtin or a constant),
--- whereas the other one represents an ongoing computation killed by lack of fuel.
-data AtomicConstraint lang meta
-  = Assign meta (TermMeta lang meta)
-  | VarEq meta meta
-  | NonInlinableSymbolEq (TermMeta lang meta) (TermMeta lang meta)
-  | NonInlinableSymbolNotEq (TermMeta lang meta) (TermMeta lang meta)
-  | OutOfFuelEq (TermMeta lang meta) (TermMeta lang meta)
-  | Native PureSMT.SExpr
-  deriving (Eq, Show)
+data ConstraintSet lang meta = ConstraintSet
+  { -- | Represents the set of productive symbolic variable assignments we know of.
+    -- Productive in the sense that we will always have at least one constructor
+    -- on the rhs.
+    csAssignments :: M.Map meta (TermMeta lang meta),
+    -- | Represents the symbolic variable equivalences we know of. Satisfies the invariant
+    -- that if @M.lookup v csmetaEq == Just v'@, then @v' < v@ (lexicographically)
+    csMetaEq :: M.Map meta meta,
```

Comment:
Is it possible to know of an equivalence between two meta-variables without knowing any assignment of them? My first instinct was that, no, it wasn't possible. But now I'm pretty sure it could be. Whatever the answer is, we should probably specify it. Also, if it is possible, then some code later looks quite fishy to me.

Reply:
In particular, the notion of [“new `v`”](https://github.com/tweag/pirouette/blob/f12b95104b8b17996228587867c85f193944b7ab/src/Pirouette/SMT/Constraints.hs#L146) in `unifyMetaWith` is a bit counter-intuitive, because it is not actually new, just without assignment. Using a union-find (which I'm in the process of doing) would make this part of the code clearer, but I need to be sure about what the semantics is *supposed to be* before making sure that I respect it.